porch/engine/pkg/engine/kuberuntime.go

```diff
@@ -0,0 +1,190 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package engine
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"math/rand"
+	"time"
+
+	v1 "github.com/GoogleContainerTools/kpt/pkg/api/kptfile/v1"
+	"github.com/GoogleContainerTools/kpt/pkg/fn"
+	"github.com/GoogleContainerTools/kpt/porch/engine/pkg/kpt"
+	"github.com/GoogleContainerTools/kpt/porch/func/evaluator"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func newKubeFunctionRuntime(cl client.Client, namespace string) (*kubeFunctionRuntime, error) {
+	return &kubeFunctionRuntime{
+		kubeClient: cl,
+		namespace:  namespace,
+	}, nil
+}
+
+type kubeFunctionRuntime struct {
+	// kubeClient is the kubernetes client
+	kubeClient client.Client
+
+	// namespace holds the namespace where the executors run
+	namespace string
+}
+
+var _ kpt.FunctionRuntime = &kubeFunctionRuntime{}
+
+func (kr *kubeFunctionRuntime) GetRunner(ctx context.Context, fn *v1.Function) (fn.FunctionRunner, error) {
+	// Create a pod
+	pod := &corev1.Pod{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "Pod",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: kr.namespace,
+			Name:      fmt.Sprintf("krm-fn-%v", randStringRunes(6)),
+			// TODO: add labels
+		},
+		Spec: corev1.PodSpec{
+			InitContainers: []corev1.Container{
+				{
+					Name:            "copy-wrapper-server",
+					Image:           "gcr.io/mengqiy-playground/wrapper-server:v0", // TODO
+					ImagePullPolicy: "Always",
+					VolumeMounts: []corev1.VolumeMount{
+						{
+							Name:      "wrapper-server",
+							MountPath: "/wrapper-server",
+						},
+					},
+				},
+			},
+			Containers: []corev1.Container{
+				{
+					Name:    "function",
+					Image:   fn.Image,
+					Command: []string{"/wrapper-server/wrapper-server"},
+					VolumeMounts: []corev1.VolumeMount{
+						{
+							Name:      "wrapper-server",
+							MountPath: "/wrapper-server",
+						},
+					},
+				},
+			},
+			Volumes: []corev1.Volume{
+				{
+					Name: "wrapper-server",
+					VolumeSource: corev1.VolumeSource{
+						EmptyDir: &corev1.EmptyDirVolumeSource{},
+					},
+				},
+			},
+		},
+	}
+	err := kr.kubeClient.Create(ctx, pod)
+	if err != nil {
+		return nil, err
+	}
+
+	return &kubeFunctionRunner{
+		ctx:   context.TODO(), // TODO: should we use the provided ctx here?
```

Comment:
TBH we shouldn't be embedding the context in the runner object; ideally Run would take a ctx